precompiles/xcm/src/lib.rs

```diff
@@ -202,57 +211,70 @@ where
                     network: None,
                     id: recipient,
-                })
+                }
             }
             BeneficiaryType::Account20 => {
                 let recipient: H160 = input.read::<Address>()?.into();
-                X1(Junction::AccountKey20 {
+                Junction::AccountKey20 {
                     network: None,
                     key: recipient.to_fixed_bytes(),
-                })
+                }
             }
         }
         .into();
 
         let is_relay = input.read::<bool>()?;
-        let parachain_id: u32 = input.read::<U256>()?.low_u32();
-        let fee_asset_item: u32 = input.read::<U256>()?.low_u32();
 
-        if fee_asset_item as usize > assets.len() {
-            return Err(revert("Bad fee index."));
-        }
+        let parachain_id: u32 = input
+            .read::<U256>()?
+            .try_into()
+            .map_err(|_| revert("error converting parachain_id, maybe value too large"))?;
 
-        // Prepare pallet-xcm call arguments
-        let dest = if is_relay {
+        let fee_item: u32 = input
+            .read::<U256>()?
+            .try_into()
+            .map_err(|_| revert("error converting fee_index, maybe value too large"))?;
+
+        let mut destination = if is_relay {
             MultiLocation::parent()
         } else {
             X1(Junction::Parachain(parachain_id)).into_exterior(1)
         };
 
-        let assets: MultiAssets = assets
+        destination
+            .push_interior(beneficiary)
+            .map_err(|_| revert("error converting destination multilocation"))?;
+
+        let assets = assets
             .iter()
             .cloned()
             .zip(amounts.iter().cloned())
             .map(Into::into)
-            .collect::<Vec<MultiAsset>>()
-            .into();
+            .collect::<Vec<MultiAsset>>();
 
         // Build call with origin.
         let origin = Some(Runtime::AddressMapping::into_account_id(
             handle.context().caller,
         ))
         .into();
-        let call = pallet_xcm::Call::<Runtime>::reserve_withdraw_assets {
-            dest: Box::new(dest.into()),
-            beneficiary: Box::new(beneficiary.into()),
-            assets: Box::new(assets.into()),
-            fee_asset_item,
+
+        // Since multiassets sorts them, we need to check whether the index is still correct,
+        // and error otherwise as there is not much we can do other than that
+        let multiassets = MultiAssets::from_sorted_and_deduplicated(assets).map_err(|_| {
+            revert("In field Assets, Provided assets either not sorted nor deduplicated")
+        })?;
+
+        let call = orml_xtokens::Call::<Runtime>::transfer_multiassets {
+            assets: Box::new(VersionedMultiAssets::V3(multiassets)),
+            fee_item,
+            dest: Box::new(VersionedMultiLocation::V3(destination)),
+            dest_weight_limit: WeightLimit::Unlimited,
```

Comment:
We have to decide about the WeightLimit to be used in these functions. I couldn't find any way to set it to anything else as xtokens decide the exact xcm instructions to be used for a transfer.

We can do a hacky solution in which we can ask users to provide us with weights in the array ` uint256[] calldata asset_amount,` but that is not very elegant.

Thoughts?

Reply:
> We can do a hacky solution in which we can ask users to provide us with weights in the array  uint256[] calldata asset_amount, but that is not very elegant.

We should not, this'll be a breaking change for existing contracts that are dependent on it.
We don't have much option here other than using `Unlimited`

Reply:
I agree, waiting for @shaunxw to comment.

Reply:
There are two options:
1. Weigh the XCM sequence used in transfer operation, and use that weight here.
2. Use `Unlimited`.

The benefit of the first approach is that it keeps the same behavior as before.
However, it's dirty since we'd be assuming how the sequence looks like.

The second solution is simple & elegant, but doesn't fully conform with previous behavior.
However, it doesn't break it which is the most important factor here.

So we should just use `Unlimited`, IMO.

Reply:
Agreed. `Unlimited` makes the most sense. We don't have to replicate the weight handling in `reserve_withdraw_assets`.